topics/data-science/tutorials/python-loops/tutorial.md

```diff
@@ -0,0 +1,551 @@
+> >     else:
+> >         print(i * 2)
+> > # But what about this line:
+> > print(len(data))
+> > ```
+> >
+> > Here this code is actually ambiguous, we don't know how indented `print(len(data))` should be. This very synthetic example lacks context, but there are three places it could be, with three different effects.
+> > The first option, no indentation, prints out the list length at the end.
+> >
+> > ```python
+> > [...]
+> >     else:
+> >         print(i * 2)
+> > print(len(data))
+> > ```
+> >
+> > The second, prints out the length once per loop iteration.
+> >
+> > ```python
+> > data = [1, 3, 5, 9]
+> > for i in data:
+> >     if i < 4:
+> >         print(i)
+> >     else:
+> >         print(i * 2)
+> >     print(len(data))
+> > ```
+> >
+> > First the `if` and `else` case resolve themselves, and then `len(data)` is printed every single loop iteration.
+> > The last case only prints out the length if `i >= 4`.
+> >
+> > ```python
+> > data = [1, 3, 5, 9]
+> > for i in data:
+> >     if i < 4:
+> >         print(i)
+> >     else:
+> >         print(i * 2)
+> >         print(len(data))
+> > ```
+> >
+> {: .solution}
+{: .question}
+
+```python
+# Check your answers here!
+data = [1, 3, 5, 9]
+for i in data:
+if i < 4:
+print(i)
+else:
+print(i * 2)
+print(len(data))
+```
+
+## Variable Naming
+
+Loop variables can be called anything, `i`, `j`, and `k` are very commong defaults due to their long history of use in other programing languages.
+As with all variables, loop variables are: Created on demand, and Meaningless; their names can be anything at all.
+
+```python
+for kitten in [2, 3, 5]:
+    print(kitten)
+```
+
+But *meaningless* is bad for variable names, and whenever possible, we should strive to pick useful, accurate variable names that help use remember what's going on:
+
+```
+for sequence in sequences:
+    print()
+for patient in clinic_patients:
+    print()
+for nucleotide in dna_sequence:
+    print()
+```
+
+## Range
+
+You can use `range` to iterate over a sequence of numbers. This is a built in function (check `help(range)`!) so it's always available even if you don't `import` anything. The range produced is non-inclusive: `range(N)` is the numbers `0` to `N-1`, so the result will be exactly the length you requested.
+
+```python
+for number in range(3):
+    print(number)
+```
+
+However in python `range` is a special type of iterable: none of the numbers are created until we need them.
+
+```python
+print(range(5))
+print(range(-3, 8)[0:4])
+```
+
+The easiest way to see what numbers are actually in there is to convert it to a `list`:
+
+```python
+print(list(range(5)))
+print(list(range(-3, 8)))
+print(list(range(0, 10, 2)))
+```
+
+## Accumulation
+
+In programming you'll often want to accumulate some values: counting things (or "accumulating"). The pattern consists of creating a variable to store your result, running a loop over some data, and in that loop, adding to the variable for your result.
+
+```python
```

Comment:
Note!. In previous python cells was there a index level bug. but not in this one